aptos-move/framework/aptos-stdlib/sources/cryptography/groth16.move

```diff
@@ -0,0 +1,96 @@
+module aptos_std::groth16 {
+    #[test_only]
+    use aptos_std::curves::{BLS12_381_G1, BLS12_381_G2, bytes_into_point, bytes_into_scalar, BLS12_381_Gt};
+    use aptos_std::curves;
+
+    struct VerifyingKey<phantom G1, phantom G2, phantom Gt> has drop {
+        alpha_g1: curves::Point<G1>,
+        beta_g2: curves::Point<G2>,
+        gamma_g2: curves::Point<G2>,
+        delta_g2: curves::Point<G2>,
+        gamma_abc_g1: vector<curves::Point<G1>>,
+    }
+
+    struct Proof<phantom G1, phantom G2, phantom Gt> has drop {
+        a: curves::Point<G1>,
+        b: curves::Point<G2>,
+        c: curves::Point<G1>,
+    }
+
+    public fun new_vk<G1,G2,Gt>(alpha_g1: curves::Point<G1>, beta_g2: curves::Point<G2>, gamma_g2: curves::Point<G2>, delta_g2: curves::Point<G2>, gamma_abc_g1: vector<curves::Point<G1>>): VerifyingKey<G1,G2,Gt> {
+        VerifyingKey {
+            alpha_g1,
+            beta_g2,
+            gamma_g2,
+            delta_g2,
+            gamma_abc_g1,
+        }
+    }
+
+    public fun new_proof<G1,G2,Gt>(a: curves::Point<G1>, b: curves::Point<G2>, c: curves::Point<G1>): Proof<G1,G2,Gt> {
+        Proof { a, b, c }
+    }
+
+    public fun verify_proof<G1,G2,Gt>(_vk: &VerifyingKey<G1,G2,Gt>, _public_inputs: &vector<curves::Scalar<G1>>, _proof: &Proof<G1,G2,Gt>): bool {
+        let gamma_abc_g1_handles = vector[];
+        let gamma_abc_g1_count = std::vector::length(&_vk.gamma_abc_g1);
+        let i = 0;
+        while (i < gamma_abc_g1_count) {
+            let item = std::vector::borrow(&_vk.gamma_abc_g1, i);
+            let handle = curves::get_point_handle(item);
+            std::vector::push_back(&mut gamma_abc_g1_handles, (handle as u8));
+            i = i + 1;
+        };
+
+        let public_input_handles: vector<u8> = vector[];
+        let public_input_count = std::vector::length(_public_inputs);
+        let i = 0;
+        while (i < public_input_count) {
+            let item = std::vector::borrow(_public_inputs, i);
+            let handle = curves::get_scalar_handle(item);
+            std::vector::push_back(&mut public_input_handles, (handle as u8));
+            i = i + 1;
+        };
+
+        verify_proof_internal(
```

Comment:
But why have a native function here? You can do the verification directly in Move, no?